Documentation/teaching/lectures/smp.rst

```diff
@@ -0,0 +1,1088 @@
+==========================
+Symmetric Multi-Processing
+==========================
+
+`View slides <smp-slides.html>`_
+
+.. slideconf::
+   :autoslides: False
+   :theme: single-level
+
+Lecture objectives:
+===================
+
+.. slide:: Symmetric Multi-Processing
+   :inline-contents: True
+   :level: 2
+
+   * Kernel Concurrency
+
+   * Atomic operations
+
+   * Spin locks
+
+   * Cache thrashing
+
+   * Optimized spin locks
+
+   * Process and Interrupt Context Synchronization
+
+   * Mutexes
+
+   * Per CPU data
+
+   * Memory Ordering and Barriers
+
+   * Read-Copy Update
+
+
+Synchronization basics
+======================
+
+Because the Linux kernel support SMP (symmetric multi-processing) but
+also because we can have concurrency between process and interrupt
+contexts it uses a set of synchronization mechanisms to achieve
+predictable results, free of race conditions.
+
+Race conditions can occur when:
+
+.. slide:: Race conditions
+   :inline-contents: True
+   :level: 2
+
+   * there are at least two execution contexts that run in "parallel":
+
+     * truly run in parallel (e.g. two system calls running on
+       different processors)
+
+     * one of the contexts can arbitrary preempt the other (e.g. an
+       interrupt preempts a system call)
+
+   * the execution contexts perform read-write accesses to shared
+     memory
+
+
+A race condition can results in different behavior when running the
+same contexts multiple times because the order and duration of the
+contexts in such a situation can affect the end result.
+
+A classical race condition example is an incorrect implementation for
+a release operation of a resource counter:
+
+.. slide:: Race condition: resource counter release
+   :inline-contents: True
+   :level: 2
+
+   .. code-block:: c
+
+      void release_resource()
+      {
+           counter--;
+
+	   if (!counter)
+               free_resource();
+      }
+
+
+A resource counter is used to keep a shared resource available until
+the last user release it but the above implementation has a race
+condition that can cause freeing the resource twice:
+
+
+.. slide:: Race condition scenario
+   :inline-contents: True
+   :level: 2
+
+   |_|
+
+   .. ditaa::
+
+                             counter is 2
+
+        Thread A                                Thread B
+
+           *
+           |
+           |
+     +---------------------+
+     |  dec counter        |   counter is 1
+     |  cEEE        	   |
+     +---------------------+
+           |
```

Comment:
What does cEEE mean?

Reply:
cEEE is a color code that fills the rectablge with gray